### PR TITLE
Complete migration over to REST API and Catalog/Engine + Build changes

### PR DESCRIPTION
**Pull request by samredai**
### Summary

This PR completes the migration to the REST API, DJ AST, and Catalog/Engine concept. Since this also required a lot of changes to the build (thanks @CircArgs for all the help!) it was hard to keep the build in a transition state where it worked along smaller piecemeal changes, hence the bigger PR. Here's a list of the changes:

- Fully migrated all yaml based logic over to the REST api
- Consolidated all node creation in tests to a single fixture function load_examples that makes API calls to a test client and can optionally be used in any given test. (Also I create all of the examples there that where in YAML before)
- Changes all Database logic to Catalog+Engine logic.
- Adds common catalog validation through change management, allowing the build to assume a common catalog for all node paths and also materializations. (**Way** faster build, linear time complexity, no more build planning)
- Changes to the build to use availabilities posted by a materialization service and instrumented logic to automatically check a build criteria method. I haven’t added the analysis yet, that logic should be discussed during the Materialization service development. (An example of the build using availability has been added to the roads example notebook)
- Reverted back to external tables being defined directly on a source node instead of as a separate entity (buyer’s remorse). Also source nodes now require a catalog and that catalog is automatically propagated down to downstream nodes.
- Transforms/dimension nodes that do multi-catalog joins are now caught during node creation (also catches cube node creation with elements from multiple dimensions). Later when we add support for multi-catalog joins we can have a setting to enable it just like postgres.
- Removed the watchdog container and repository logic. As discussed a few syncs back, we can move these to a client library for users to utilize in a distributed way (many different repos) which all post back to a central DJ server. We can also add the YAML parsing logic there that was removed here for users who want to use that pattern. (Probably a good time to start on the client library)

### Test Plan

Create and build all examples through the API (basic, dbt, and roads), run all notebook examples, run `make check` and `make test`.

- [x] PR has an associated issue: #382 #298 #271 #255 #121 #305 #301
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### Deployment Plan

The only remaining API endpoints planned are the ones around registering functions. Once that's added, I think we may be ready to stabilize the REST API spec (which we should keep stable within major DJ releases).

### CODE REVIEW

**Review by CircArgs — APPROVED** [review]

File: Dockerfile
Comment (shangyian) [inline Dockerfile:10]: Do we want to do this here or keep track of these in a separate requirements file?

File: dj/models/table.py
Comment (shangyian) [inline dj/models/table.py:56]: Just remove this?

File: docker-compose.yml
Comment (shangyian) [inline docker-compose.yml:29]: Won't removing `extra_hosts` cause issues with running the query service?

File: dj/api/graphql/column.py
Comment (shangyian) [inline dj/api/graphql/column.py:1]: These don't get added back right? Is the idea that we can add the graphql API later, after the existing rest API is more settled?

File: dj/api/nodes.py
Comment (shangyian) [inline dj/api/nodes.py:553]: Not a blocker, but just flagging that we should add this again so that we can infer columns for loaded tables (in a follow-up PR)

**Review by shangyian — APPROVED** [review]
Changes looks good! I left a few minor comments - mostly about follow-ups, nothing blocking.

File: dj/api/graphql/column.py
Comment (samredai) [reply dj/api/graphql/column.py:1]: Yeah we needed to thoroughly comb through the graphql API anyway so it seemed much easier to add this back and confirm parity when we're locked into an openapi spec for an entire major version (cc: @CircArgs).

Comment (shangyian) [reply dj/api/graphql/column.py:1]: Makes sense!

File: Dockerfile
Comment (samredai) [reply Dockerfile:10]: Let me try moving it to the test requirements. I believe the only place it's used is in the pre-commit runs to make sure the example notebooks are all formatted nicely.

File: dj/api/nodes.py
Comment (samredai) [reply dj/api/nodes.py:553]: Another great catch! Let me move this over to the create node logic for source nodes since the table info is back on their definitions now.

File: Dockerfile
Comment (samredai) [reply Dockerfile:10]: Actually, pre-commit handles it's own requirements, I just went ahead and removed this.

File: dj/models/table.py
Comment (samredai) [reply dj/models/table.py:56]: Done

File: dj/api/helpers.py
Comment (shangyian) [inline dj/api/helpers.py:125]: Does this need to also include the engine? As in, now instead of building the query for a database, we're building the query for an engine. Or is the idea that this is just a generic query AST, which would then need to be translated to engine-specific SQL by the query engine?

File: dj/construction/utils.py
Comment (agorajek) [inline dj/construction/utils.py:108]: If someone wanted to have `_foo_` we would make it `foo` ?

File: dj/models/node.py
Comment (agorajek) [inline dj/models/node.py:265]: You don't like the `# TODO: ...` marker?

File: tests/conftest.py
Comment (agorajek) [inline tests/conftest.py:4]: lol, @samredai please start writing smaller PRs. I bet in Java this lint error would come up in the Hello World app.

File: tests/conftest.py
Comment (agorajek) [inline tests/conftest.py:113]: YAML to JSON... yuck! :)

File: tests/construction/dj_query_test.py
Comment (agorajek) [inline tests/construction/dj_query_test.py:30]: If we show these `_DOT_`s to some users, they'll think we are insane ;)

File: tests/errors_test.py
Comment (agorajek) [inline tests/errors_test.py:46]: This looks like `assert foo == foo` to me. What am I missing?

File: tests/sql/parsing/queries/case_when_null.sql
Comment (agorajek) [inline tests/sql/parsing/queries/case_when_null.sql:6]: `IS NOT NULL` is more SQL-ish

File: tests/sql/parsing/test_ast.py
Comment (agorajek) [inline tests/sql/parsing/test_ast.py:503]: Looks like collateral damage from a `codemod "_to_select" "to_select"` ?

**Review by agorajek — APPROVED** [review]
This PR feels like a new era for DJ. As much as I would like to hold on to my precious little things, I know DJ is becoming and adult and needs to move on ;)

Great work @samredai !

Just a few comments in line, some of which you may want to address before merge.

File: alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py
Comment (samredai) [reply alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py:289]: This is actually autogenerated by alembic from the sqlmodel `CubeRelationship` class. Since cube definitions are just node definitions, they share the `node` table and this is just a two column lookup table used as a "link model". Would a better name be something like "CubeLookup" or "CubeLink"?

File: dj/models/node.py
Comment (samredai) [reply dj/models/node.py:265]: I do but we have a pre-commit check which will fail if it finds any to-do's. I know this is basically an attempt at circumventing that but I'm expecting to follow-up pretty quickly with an implementation of this function. (famous last words, I know...)

File: alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py
Comment (betodealmeida) [reply alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py:289]: @samredai you can override the default table names generated for models using the `__tablename__` attribute (something like that).

File: tests/conftest.py
Comment (samredai) [reply tests/conftest.py:4]: Haha, I did feel awkward putting that there. It's because of the long list of examples I added to the conftest. Thanks for keeping me honest here, I moved them to an examples.py file and now both files are under the limit, allowing me to remove this. 🥳

File: tests/conftest.py
Comment (samredai) [reply tests/conftest.py:113]: Don't worry, the DJ python client will have the greatest YAML to JSON converter the world has ever known! 😂

File: tests/errors_test.py
Comment (samredai) [reply tests/errors_test.py:46]: This is to check that the `__eq__` method is working properly...and to get coverage on those lines :D

File: dj/api/nodes.py
Comment (samredai) [reply dj/api/nodes.py]: Good idea! The requirements now are that each node have a catalog but you're right, it's good to explicitly to catch any strange behaviors. Added a `< 1` check.

File: dj/api/nodes.py
Comment (samredai) [reply dj/api/nodes.py]: Fixed 👍

File: tests/sql/parsing/test_ast.py
Comment (samredai) [reply tests/sql/parsing/test_ast.py]: You're exactly right, that was the exact find and replace that I did. 😅 (fixed it!)

File: tests/sql/parsing/queries/case_when_null.sql
Comment (samredai) [reply tests/sql/parsing/queries/case_when_null.sql:6]: I think this was due to a small bug in the build. I believe @CircArgs is planning a fast follow-up to fix this.

File: tests/construction/dj_query_test.py
Comment (samredai) [reply tests/construction/dj_query_test.py:30]: Agreed, I'll add this to the sync to discuss the best syntax for these system generated aliases.

File: alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py
Comment (samredai) [reply alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py:289]: Nice! Alright I'll keep the classname the same and change the table name to "cube"

File: dj/construction/utils.py
Comment (samredai) [reply dj/construction/utils.py:108]: Hmm...good point. I think you're right, this would be a problem. @CircArgs can you confirm?

File: dj/api/data.py
Comment (samredai) [reply dj/api/data.py]: Fixed

File: alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py
Comment (samredai) [reply alembic/versions/2023_03_10_0331-143a068da0ed_initial_migration.py]: That worked nicely, thanks @betodealmeida!

File: alembic/versions/2023_03_12_0158-e3fc8f2f003f_initial_migration.py
Comment (betodealmeida) [inline alembic/versions/2023_03_12_0158-e3fc8f2f003f_initial_migration.py:24]: Nit: can we remove these so we know the auto-generated migration was revised? Thanks!

File: alembic/versions/2023_03_12_0158-e3fc8f2f003f_initial_migration.py
Comment (betodealmeida) [inline alembic/versions/2023_03_12_0158-e3fc8f2f003f_initial_migration.py:422]: Ditto.

File: dj/construction/utils.py
Comment (samredai) [reply dj/construction/utils.py:108]: Since this is the only remaining comment, I'm going to merge this to unblock the work in a few people's forks so that they can rebase.

File: dj/models/node.py
Comment (betodealmeida) [inline dj/models/node.py:497]: If these are optional should we make the default `None`?

File: dj/models/table.py
Comment (betodealmeida) [reply dj/models/table.py:56]: It's still here? 😛

File: tests/api/data_test.py
Comment (betodealmeida) [inline tests/api/data_test.py:22]: This is a bit confusing, since in other tests we need to explicitly call `load_examples`, but here the `client` fixture calls it automatically. It would be better to be consistent in the fixtures so we don't have surprises. Maybe call this fixture `client_with_examples`?

**Review by betodealmeida — APPROVED** [review]
This is great! I left a comment on the `client` fixture being inconsistent, otherwise looks good.

File: dj/models/table.py
Comment (samredai) [reply dj/models/table.py:56]: Thanks @betodealmeida!
image

File: alembic/versions/2023_03_12_0158-e3fc8f2f003f_initial_migration.py
Comment (samredai) [reply alembic/versions/2023_03_12_0158-e3fc8f2f003f_initial_migration.py]: Done!

File: dj/models/node.py
Comment (samredai) [reply dj/models/node.py]: Good idea, updated 👍

File: dj/models/table.py
Comment (samredai) [reply dj/models/table.py]: Ok, now it's really really removed. lol

File: tests/api/data_test.py
Comment (samredai) [reply tests/api/data_test.py:22]: This is a great catch, I like the `client_with_examples`, updated this pattern to use that.

**Pull request merged**
